File: src/iac_code/tools/cloud/aliyun/aliyun_doc_search.py
"""AliyunDocSearch - searches Alibaba Cloud documentation."""

from __future__ import annotations

from typing import Any

import httpx

from iac_code.i18n import _
from iac_code.services.telemetry import log_event
from iac_code.services.telemetry.names import Events
from iac_code.tools.base import Tool, ToolContext, ToolResult

_SEARCH_URL = "https://help.aliyun.com/help/json/search.json"
_TIMEOUT = 10
_PAGE_SIZE = 10


class AliyunDocSearch(Tool):
    """Tool for searching Alibaba Cloud documentation."""

    @property
    def name(self) -> str:
        return "aliyun_doc_search"

    @property
    def description(self) -> str:
        return (
            "Search Alibaba Cloud documentation. Returns document titles, summaries and links. "
            "Use category_id=28850 to limit results to ROS product docs."
        )

    @property
    def input_schema(self) -> dict[str, Any]:
        return {
            "type": "object",
            "properties": {
                "keywords": {
                    "type": "string",
                    "description": "Search keywords",
                },
                "category_id": {
                    "type": "integer",
                    "description": "Product category ID, e.g. 28850 for ROS. Omit to search all products.",
                },
            },
            "required": ["keywords"],
        }

    def is_read_only(self, input: dict | None = None) -> bool:
        return True

    def is_destructive(self, input: dict | None = None) -> bool:
        return False

    def user_facing_name(self, input: dict | None = None) -> str:
        return _("DocSearch")

    def render_tool_use_message(self, input: dict, *, verbose: bool = False) -> str | None:
        return input.get("keywords")

    def render_tool_result_message(self, output: str, *, is_error: bool = False, verbose: bool = False) -> str | None:
        if is_error:
            return output
        if hasattr(self, "_last_summary"):
            return self._last_summary
        return self._summary_from_persisted_output(output)

    @staticmethod
    def _summary_from_persisted_output(output: str) -> str | None:
        for line in reversed(output.splitlines()):
            text = line.strip()
            if not text or "web_fetch" in text:
                continue
            lower = text.lower()
            if "found" in lower and "document" in lower:
                return text
            if "no documents found" in lower:
                return text
            if "文档" in text and ("找到" in text or "共" in text):
                return text
        return None

    def get_activity_description(self, input: dict | None = None) -> str | None:
        if input:
            keywords = input.get("keywords", "")
            return _("Searching docs for {keywords}...").format(keywords=keywords)
        return _("Searching docs...")

    async def execute(self, *, tool_input: dict[str, Any], context: ToolContext) -> ToolResult:
        keywords = tool_input.get("keywords", "").strip()
        if not keywords:
            return ToolResult.error(_("keywords cannot be empty."))

        category_id = tool_input.get("category_id")

        params: dict[str, Any] = {
            "keywords": keywords,
            "topics": "DOCUMENT,PRODUCT",
            "language": "zh",
            "website": "cn",
            "pageSize": _PAGE_SIZE,
            "pageNum": 1,
        }
        if category_id is not None:
            params["categoryId"] = category_id

        try:
            async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
                response = await client.get(_SEARCH_URL, params=params)
                response.raise_for_status()
        except httpx.HTTPStatusError as e:
            return ToolResult.error(_("HTTP error {status} when searching docs.").format(status=e.response.status_code))
        except httpx.HTTPError as e:
            return ToolResult.error(_("Failed to search docs: {error}").format(error=str(e)))

        try:
            data = response.json()
        except Exception:
            return ToolResult.error(_("Failed to parse search response as JSON."))

        if not data.get("success"):
            return ToolResult.error(_("Search API returned failure."))

        documents = data.get("data", {}).get("documents", {})
        items = documents.get("data", [])
        total = documents.get("totalCount", 0)

        # Emit doc search event
        category_str = str(category_id) if category_id is not None else None
        log_event(
            Events.DOC_SEARCHED,
            {
                "doc_source": "aliyun_ros_api",
                "search_category": category_str,
                "result_count": len(items),
                "outcome": "success" if items else "empty",
            },
        )

        if not items:
            self._last_summary = _("No documents found")
            return ToolResult.success(_("No documents found for keywords: {keywords}").format(keywords=keywords))

        lines: list[str] = []
        for i, item in enumerate(items, 1):
            title = item.get("title", "")
            content = item.get("content", "")
            url = item.get("url", "")
            lines.append(f"{i}. {title}")
            if content:
                lines.append(f"   {content}")
            if url:
                lines.append(f"   Link: {url}")
            lines.append("")

        count = len(items)
        self._last_summary = _("Found {count} documents (total {total})").format(count=count, total=total)
        lines.append(self._last_summary)
        lines.append(_("Use web_fetch tool to read full document content if needed."))

        return ToolResult.success("\n".join(lines))

File: iac-code-rs/crates/iac-code-tools/src/aliyun_doc_search.rs
use iac_code_protocol::json::{self, JsonValue};

use std::path::PathBuf;

use iac_code_config::cloud_credentials::AliyunCredential;

use crate::{
    AliyunApiTool, RosStackInstancesTool, RosStackTool, Tool, ToolContext, ToolRegistry, ToolResult,
};

mod input;
mod response;
mod schema;

use input::{build_search_params, string_field};
use response::format_search_response;
use schema::input_schema;

const SEARCH_URL: &str = "https://help.aliyun.com/help/json/search.json";
const PAGE_SIZE: usize = 10;

const ALIYUN_TOOL_NAMES: &[&str] = &[
    "aliyun_api",
    "aliyun_doc_search",
    "ros_stack",
    "ros_stack_instances",
];

#[derive(Clone, Debug)]
pub struct AliyunDocSearchTool {
    search_url: String,
}

impl AliyunDocSearchTool {
    pub fn new() -> Self {
        Self {
            search_url: SEARCH_URL.into(),
        }
    }

    pub fn with_search_url(mut self, search_url: impl Into<String>) -> Self {
        self.search_url = search_url.into();
        self
    }
}

impl Default for AliyunDocSearchTool {
    fn default() -> Self {
        Self::new()
    }
}

impl Tool for AliyunDocSearchTool {
    fn name(&self) -> &str {
        "aliyun_doc_search"
    }

    fn description(&self) -> &str {
        "Search Alibaba Cloud documentation. Returns document titles, summaries and links. Use category_id=28850 to limit results to ROS product docs."
    }

    fn input_schema(&self) -> JsonValue {
        input_schema()
    }

    fn validate_input(&self, input: &JsonValue) -> Result<(), String> {
        match string_field(input, "keywords") {
            Some(_) => Ok(()),
            None => Err("missing required field 'keywords'".into()),
        }
    }

    fn execute(&self, input: &JsonValue, _context: &ToolContext) -> ToolResult {
        let keywords = string_field(input, "keywords").unwrap_or_default().trim();
        if keywords.is_empty() {
            return ToolResult::error("keywords cannot be empty.");
        }

        let params = build_search_params(input, keywords, PAGE_SIZE);

        let client = reqwest::blocking::Client::builder()
            .timeout(std::time::Duration::from_secs(10))
            .build();
        let Ok(client) = client else {
            return ToolResult::error("Failed to search docs: failed to build HTTP client");
        };
        let response = match client.get(&self.search_url).query(&params).send() {
            Ok(response) => response,
            Err(error) => return ToolResult::error(format!("Failed to search docs: {error}")),
        };
        let status = response.status();
        if !status.is_success() {
            return ToolResult::error(format!(
                "HTTP error {} when searching docs.",
                status.as_u16()
            ));
        }
        let body = match response.text() {
            Ok(body) => body,
            Err(error) => return ToolResult::error(format!("Failed to search docs: {error}")),
        };
        let data = match json::parse(&body) {
            Ok(data) => data,
            Err(_) => return ToolResult::error("Failed to parse search response as JSON."),
        };
        if bool_field(&data, "success") != Some(true) {
            return ToolResult::error("Search API returned failure.");
        }

        ToolResult::success(format_search_response(&data, keywords))
    }

    fn is_read_only(&self, _input: &JsonValue) -> bool {
        true
    }

    fn user_facing_name(&self, _input: &JsonValue) -> String {
        "DocSearch".into()
    }
}

pub fn register_cloud_tools(
    registry: &mut ToolRegistry,
    aliyun_credential: Option<AliyunCredential>,
) {
    register_cloud_tools_with_cloud_credentials_path(registry, aliyun_credential, None);
}

pub fn register_cloud_tools_with_cloud_credentials_path(
    registry: &mut ToolRegistry,
    aliyun_credential: Option<AliyunCredential>,
    cloud_credentials_path: Option<PathBuf>,
) {
    for tool_name in ALIYUN_TOOL_NAMES {
        registry.unregister(tool_name);
    }
    if let Some(credential) = aliyun_credential {
        let mut api_tool = AliyunApiTool::new(Some(credential.clone()));
        let mut ros_stack_tool = RosStackTool::new(Some(credential.clone()));
        let mut ros_stack_instances_tool = RosStackInstancesTool::new(Some(credential.clone()));
        if let Some(path) = cloud_credentials_path {
            api_tool = api_tool.with_cloud_credentials_path(path.clone());
            ros_stack_tool = ros_stack_tool.with_cloud_credentials_path(path.clone());
            ros_stack_instances_tool = ros_stack_instances_tool.with_cloud_credentials_path(path);
        }
        registry.register(Box::new(api_tool));
        registry.register(Box::new(AliyunDocSearchTool::new()));
        registry.register(Box::new(ros_stack_tool));
        registry.register(Box::new(ros_stack_instances_tool));
    }
}

fn bool_field(input: &JsonValue, field: &str) -> Option<bool> {
    let JsonValue::Object(fields) = input else {
        return None;
    };
    match fields.get(field) {
        Some(JsonValue::Bool(value)) => Some(*value),
        _ => None,
    }
}
